File: composeApp/src/commonMain/kotlin/com/aus/notelikeus/domain/reminder/ReminderTime.kt
package com.aus.notelikeus.domain.reminder

import com.aus.notelikeus.util.DateUtils

/**
 * The instants a reminder can be set to, and the conversions a date/time picker needs.
 *
 * This exists as a domain object rather than as arithmetic inside the dialog because the two
 * conversions below are the ones that go wrong quietly. Both are about the same hazard: a civil
 * date and an instant are different things, and every reminder bug this project has had came from
 * treating one as the other.
 */
object ReminderTime {

    /** One of the quick choices offered alongside picking an exact moment. */
    enum class Preset { IN_ONE_HOUR, TOMORROW_MORNING, NEXT_WEEK }

    const val ONE_HOUR_MS: Long = 60 * 60 * 1000L

    fun resolve(preset: Preset): Long = when (preset) {
        Preset.IN_ONE_HOUR -> DateUtils.currentTimeMillis() + ONE_HOUR_MS
        Preset.TOMORROW_MORNING -> DateUtils.getTomorrowMorning()
        Preset.NEXT_WEEK -> DateUtils.getNextWeek()
    }

    /**
     * Where the picker should open: the existing reminder, or the next whole hour.
     *
     * Returns the *date picker's* value in the units Material3 uses — UTC midnight — alongside the
     * local hour and minute, so the two halves of the dialog agree about which day is selected.
     */
    fun initialPickerState(existingReminder: Long?): PickerState {
        val anchor = existingReminder ?: nextWholeHour(DateUtils.currentTimeMillis())
        val fields = DateUtils.localDateTimeFields(anchor)
        return PickerState(
            dateUtcMillis = civilDateToUtcMillis(fields.year, fields.month, fields.day),
            hour = fields.hour,
            minute = fields.minute,
        )
    }

    /**
     * The instant a picked civil date and time of day names, in the user's own zone.
     *
     * [datePickerUtcMillis] is what Material3's `DatePickerState.selectedDateMillis` reports:
     * midnight **UTC** on the selected civil date. Handing that value straight to a local-calendar
     * conversion is a day-boundary bug waiting for the first user west of UTC — their
     * `2026-07-08T00:00Z` reads as the evening of July 7 locally, and the reminder lands a day
     * early. So the civil date is recovered from it in UTC (exact: the value is UTC midnight by
     * contract, no zone involved), and only then handed to the platform's local-midnight
     * conversion.
     *
     * Returns null for a date that does not exist. A local time that does not exist — the hour a
     * spring-forward DST change skips — resolves forward to the first instant that does, which is
     * what [DateUtils.combineDateAndTime] already does for typed search dates.
     */
    fun exactReminder(datePickerUtcMillis: Long, hour: Int, minute: Int): Long? {
        if (hour !in 0..23 || minute !in 0..59) return null
        val civil = utcMillisToCivilDate(datePickerUtcMillis)
        val localMidnight = DateUtils.startOfDay(civil.year, civil.month, civil.day) ?: return null
        return DateUtils.combineDateAndTime(localMidnight, hour, minute)
    }

    /** The next whole hour after [now], the default a picker opens on with nothing set. */
    fun nextWholeHour(now: Long): Long {
        val fields = DateUtils.localDateTimeFields(now)
        val midnight = DateUtils.startOfDay(now)
        // hour + 1 rather than arithmetic on `now`: combineDateAndTime normalises 24 onto the next
        // day through the calendar, which is also what keeps this right across a DST change.
        return DateUtils.combineDateAndTime(midnight, fields.hour + 1, 0)
    }

    data class PickerState(val dateUtcMillis: Long, val hour: Int, val minute: Int)

    data class CivilDate(val year: Int, val month: Int, val day: Int)

    /**
     * The civil date of a UTC-midnight instant.
     *
     * Pure integer arithmetic on days since the epoch (Howard Hinnant's `civil_from_days`), so it
     * is exact for every date and involves no time zone at all — which is the point: the input is
     * defined in UTC and must not be re-interpreted anywhere else.
     */
    fun utcMillisToCivilDate(utcMillis: Long): CivilDate {
        val days = utcMillis.floorDiv(86_400_000L)
        val z = days + 719_468L
        val era = z.floorDiv(146_097L)
        val doe = z - era * 146_097L
        val yoe = (doe - doe / 1_460 + doe / 36_524 - doe / 146_096) / 365
        val y = yoe + era * 400
        val doy = doe - (365 * yoe + yoe / 4 - yoe / 100)
        val mp = (5 * doy + 2) / 153
        val d = doy - (153 * mp + 2) / 5 + 1
        val m = if (mp < 10) mp + 3 else mp - 9
        return CivilDate(
            year = (if (m <= 2) y + 1 else y).toInt(),
            month = m.toInt(),
            day = d.toInt(),
        )
    }

    /** Inverse of [utcMillisToCivilDate]: midnight UTC on the given civil date. */
    fun civilDateToUtcMillis(year: Int, month: Int, day: Int): Long {
        val y = (if (month <= 2) year - 1 else year).toLong()
        val era = y.floorDiv(400L)
        val yoe = y - era * 400
        val mp = if (month > 2) month - 3 else month + 9
        val doy = (153L * mp + 2) / 5 + day - 1
        val doe = yoe * 365 + yoe / 4 - yoe / 100 + doy
        return (era * 146_097L + doe - 719_468L) * 86_400_000L
    }
}

File: web/src/lib/reminders/reminderTime.ts
/**
 * The instants a reminder can be set to, and the conversions the picker needs.
 *
 * The Kotlin counterpart is `ReminderTime.kt`, and the presets below resolve to the same moments
 * its `DateUtils` does — `getTomorrowMorning` is 09:00 local tomorrow, `getNextWeek` is the same
 * time of day seven days out. Keeping the arithmetic here, rather than inline in the dialog, is
 * what lets the time-zone and DST behaviour be tested without a browser.
 *
 * Every function works in the **viewer's local zone**. That is the whole hazard: a
 * `datetime-local` input speaks local wall-clock time and `Date.prototype.toISOString` speaks
 * UTC, and mixing them shifts a reminder by the zone offset.
 */

export const ONE_HOUR_MS = 60 * 60 * 1000;

export type ReminderPresetId = 'in-one-hour' | 'tomorrow-morning' | 'next-week';

export interface ReminderPreset {
  id: ReminderPresetId;
  label: string;
  at: number;
}

/** The next whole hour after `now`, in local time. */
export function nextWholeHour(now: number = Date.now()): number {
  const date = new Date(now);
  date.setMinutes(0, 0, 0);
  date.setHours(date.getHours() + 1);
  return date.getTime();
}

/** 09:00 local time tomorrow. Mirrors Kotlin `DateUtils.getTomorrowMorning`. */
export function tomorrowMorning(now: number = Date.now()): number {
  const date = new Date(now);
  date.setDate(date.getDate() + 1);
  date.setHours(9, 0, 0, 0);
  return date.getTime();
}

/**
 * The same time of day, seven days out. Mirrors Kotlin `DateUtils.getNextWeek`, which adds a
 * calendar week rather than `7 * 86_400_000` — across a DST change those differ by an hour, and
 * the calendar answer is the one a user means by "next week".
 */
export function nextWeek(now: number = Date.now()): number {
  const date = new Date(now);
  date.setDate(date.getDate() + 7);
  return date.getTime();
}

export function reminderPresets(now: number = Date.now()): ReminderPreset[] {
  return [
    { id: 'in-one-hour', label: 'In 1 hour', at: now + ONE_HOUR_MS },
    { id: 'tomorrow-morning', label: 'Tomorrow 9:00', at: tomorrowMorning(now) },
    { id: 'next-week', label: 'Next week', at: nextWeek(now) },
  ];
}

function pad(value: number): string {
  return String(value).padStart(2, '0');
}

/**
 * `timestamp` as the `YYYY-MM-DDTHH:mm` string a `datetime-local` input expects.
 *
 * Built from the local getters rather than from `toISOString().slice(0, 16)`. That shortcut is
 * correct only at UTC+0: everywhere else it hands the input a wall-clock time offset by the
 * viewer's zone, so the picker opens on the wrong hour — and west of UTC, on the wrong day.
 */
export function toLocalDateTimeInputValue(timestamp: number): string {
  const date = new Date(timestamp);
  return (
    `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}` +
    `T${pad(date.getHours())}:${pad(date.getMinutes())}`
  );
}

/**
 * The instant a `datetime-local` value names, in the viewer's own zone.
 *
 * `new Date('2026-07-08T09:30')` — no trailing `Z`, no offset — is parsed as local time by every
 * current engine, which is what this input means. Returns null for a value the picker can leave
 * behind: empty, or half-typed.
 *
 * A wall-clock time a spring-forward change skips resolves to the first instant that does exist,
 * matching the Kotlin path's `combineDateAndTime`.
 */
export function fromLocalDateTimeInputValue(value: string): number | null {
  if (!/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}(:\d{2})?$/.test(value)) return null;
  const parsed = new Date(value).getTime();
  return Number.isFinite(parsed) ? parsed : null;
}

/** What the picker should open on: the existing reminder, or the next whole hour. */
export function initialReminderInputValue(
  existingReminder: number | null,
  now: number = Date.now(),
): string {
  return toLocalDateTimeInputValue(existingReminder ?? nextWholeHour(now));
}
